<!DOCTYPE html>
<html>
<head>
    <link href='http://fonts.googleapis.com/css?family=Inconsolata|Overlock+SC|Yanone+Kaffeesatz:400,300|Pathway+Gothic+One|Unica+One' rel='stylesheet' type='text/css'>
    <title>Welcome!</title>
    <style>
        
        h1 {
            font-size: 40px;
            font-family: 'Yanone Kaffeesatz', sans-serif;
            font-weight:normal;
            text-transform:uppercase;
            color: #4E4F4B;
        }
  
        h3 {
            font-family: 'Yanone Kaffeesatz', sans-serif;
            color: #D8A355;
            font-size: 24px;
            font-weight:normal;
        }
    
        h5 {
            font-family: 'Yanone Kaffeesatz', sans-serif;
            color: #D8A355;
            font-size: 21px;
            font-weight:normal;
        }
        
        h4 {
            font-size: 24px;
            font-family: 'Yanone Kaffeesatz', sans-serif;
            color: #FCC48C;
            font-weight:normal;
        }
            
        p { 
            font-size: 14px;
            font-weight: 400;
            font-family: 'Arial', sans-serif;
            color: #4E4F4B;
            margin-right: 30px;
        }
        
        ultxt {
            list-style-type:none;
            font-size:18px;
            font-family: 'Yanone Kaffeesatz', sans-serif;
            margin:0;
            padding-left:50px;
            overflow:hidden            
        }
        
        litxt {
            text-transform:uppercase;
            display:inline;
            padding:5px;
        }
        
        ulh {
            list-style-type:none;
            font-size:24px;
            font-family: 'Yanone Kaffeesatz', sans-serif;
            padding:5px;
            margin:0px;
            
        }
        
        lih {
            text-transform:uppercase;
            display:inline;
            padding:5px;
            margin:0px;
   
        }
        
        ul { 
            font-size: 14px;
            font-weight: 400;
            font-family: 'Arial', sans-serif;
            color: #4E4F4B;
        }
        
        li {
            font-size: 14px;
            font-weight: 400;
            font-family: 'Arial', sans-serif;
            color: #4E4F4B;
        }
        
        ultxt2 {
            list-style-type:none;
            font-size:16px;
            font-family: 'Yanone Kaffeesatz', sans-serif;
            margin:0;
            padding-left:65px;
            overflow:hidden            
        }
        
        litxt2 {
            text-transform:uppercase;
            display:inline;
            padding:5px;
        }
        
        #color {
            background-color:#BA522B;
            padding:8px;
        }
    </style>
</head>
<body>
    
    <br />
    <br />
    <br />
    <br />  
    <a name="top"></a>
    <h1> Mollee Jain - SURI 2013 </h1>
    <!p style="border-style: dotted; border-width: 0px; border-top-width: 5px; margin-left: 0px"> <!/p> 
<! possible colors:
    #E6AC27 - orange  #BA522B - red #4E4F4B - charcoal #80BCA3 - turq>    
<!beginning of 'About Me' section>
    <!future div style:>
    <!div style="text-align:left; border-style: solid; border-color:#4E4F4B ;border-width: 2px ;background-color:white; font-family:'Inconsolata'; padding 5px; font-size: 12px; height:200px; overflow:auto">

<div id="color">


    <ulh>
    <lih>|<a style="color:#ECE5CE" href="#aboutme">About Me</a>  |</lih>
    <lih>|<a style="color:#ECE5CE" href="#working">What I Have Been Working On</a>|</lih>
    <lih>|<a style="color:#ECE5CE" href="#useful">Useful Links</a>|</lih>
    <lih>|<a style="color:#ECE5CE" href="#contact">Contact Me</a>|</lih>
    </ulh>
</div>
    <p style="border-style: dotted; border-width: 0px; border-top-width: 5px; margin-left: 0px; margin-right:0px"> </p>
    <a name="aboutme"></a>    
    <h3> About Me: </h3>
    <p> My name is Mollee Jain and I am a rising sophomore at Wellesley College this fall.  Though my undergraduate studies are on the East Coast, my roots are right here in San Diego.  Due to the fantastic oppurtunites that the Scripps Resesarch Institute provides, I get to combine my interests in biology and computer science to learn about the world of bioinformatics at STSI under Dr. Carland. I look forward to spending the summer working with professionals! </p>
    <h5><a style="color:#BA522B; font-size:16px" href="#top">Top</a></h5>
    <p style="border-style: dotted; border-width: 0px; border-top-width: 5px; margin-left: 0px; margin-right:0px"> </p>
     

<!beginning of what I have been working on section>    
    <a name="working"></a> 
    <h3> What I Have Been Working On: </h3>
    
    <ultxt>
        <litxt>|<a style="color: #BA522B" href="#program">Programming</a>|</litxt>
        <litxt>|<a style="color: #BA522B" href="#conky">Using Conky</a>|</litxt>
        <litxt>|<a style="color: #BA522B" href="#novoalign">Running Novoalign</a>|</litxt>
        <litxt>|<a style="color: #BA522B" href="#chisq">Running a Chi-Squared Test in Python</a>|</litxt>
    </ultxt>
    
        <a name="program"></a>
        <!Programming>
        <h4> Programming: </h4>
         <p> Due to the vast number of tutorials of coding languages on the internet, I was able to find some that I could follow and understand readily to expand my knowledge of programming. Some of these tutorials include:</p>
        <p> Python: <a href="http://learnpythonthehardway.org/" style= "color: #80BCA3" target="_blank"> Learn Python The Hard Way</a></p>
        <p> PHP: <a href="http://www.html.net/tutorials/php/" style="color: #80BCA3" target="_blank"> PHP Tutorial </a> </p>
        <p> HTML and CSS: <a href="http://www.htmldog.com/" style= "color: #80BCA3" target="_blank"> HTML Dog</a></p>
        <p> CodeAcademy: <a href="http://www.codecademy.com/" style= "color: #80BCA3" target="_blank"> CodeAcademy</a> [This site provides free tutorials and practice for a multitude of programming languages!]</p>
        <p> [Note: I used <a href="http://hilite.me/" style="color:#80BCA3" target="_blank"> hilite.me </a> to format my source code in an HTML format.] </p>
 
        <!Using Conky>
        <a name="conky"></a>
        <h4> Using Conky: </h4>
        <ultxt2>
        <litxt2>|<a style="color: #BA522B" href="#setupconky">Setup</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#configureconky">Configure</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#qstatconky">qstat Command</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#myconky">My .conkyrc file</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#nextconky">What's Next?</a>|</litxt2>
    </ultxt2>
        
        <p> Conky is system monitor for your desktop that has the ability to display information in a clear and concise way. There are many different documentations for setting up Conky, and I used a combination of them to configure the Conky that I use for my desktop. </p>
        <p> Note: Conky works on Linux, FreeBSD and OpenBSD platforms - I worked on a Linux Mint platform. There are alternatives like GeekTool and Samurize for Macs, and Rainmeter and Fences for Windows.) </p>
        
        <!explaining how I created the Conky I have on my desktop>
            <a name="setupconky"></a>
            <h5> Setup: </h5>
            <p> To set up Conky with the "conky-colors" configuration, I almost entirely followed <a href="http://www.reazul.net/how-to-install-and-configure-conky-in-linux-mint/" style="color: #80BCA3" target="_blank"> this tutorial </a>.  This comprhensively combines many of the steps I took to create my configuration. If you want a different configuration, there are many, many different tutorials on the internet.  DeviantArt and NoobsLab are two websites that have a plethora of conky styles. <p>
            <p> Once you have your .conkyrc file open, try to understand the text.  Check this <a href="http://conky.sourceforge.net/docs.html" style="color:#80BCA3" target="_blank"> conky manual page</a>. It gives many definitions for the variables found in the .conkyrc file.</p>
            <a name="configureconky"></a>
            <h5> Configure: </h5>
            <p> To get the configuration I have, I changed some of the variables, shown below.  Feel free to mess around with them!</p>
            <p style="font-family:'Inconsolata', sans-serif;font-size: 13px">
            <br />own_window_transparent yes
            <br />#####[this made the window transparent]
            <br />#own_window_type override
            <br />#####[this allowed my window to stay on my desktop after I changed the .conkyrc file. Otherwise, the window disappears.]
            <br />gap_x 15
            <br />gap_y 0
            <br />######[this reduces the gap between the screen and the text]
            </p>
            <a name="qstatconky"></a>
            <h5> qstat Command </h5>
            <p> Once I was satisfied with how conky looked, I worked on creating a new section in it that reflected the 'qstat' function of what was occuring in a remote host.  The 'qstat' function allows one to see what jobs are currently running.  To create this, I inserted the following code into my .conkyrc file [note: the names (ex. remote@host.edu) should be changed to reflect your specific remote hosts and usernames, etc].</p>
            <p style="font-family:'Inconsolata', sans-serif;font-size: 13px">
            <br /> Remote Host Name ${hr 2}
            <br /> ${execi 600 ssh user@remotehost.edu 'uptime'}
            <br /> ${execi 600 ssh user@remotehost.edu 'qstat | grep username | gawk "{ print \$1 \": \" \$3 \"\t\" \"State: \" \$5 }"'}
            </p>
            <p> The number after 'execi' tells you how after how many seconds the command repeats.  In this case, this command would run every 10 minutes ('600'). By using 'grep username', you will only see jobs that are under your name (this is especially useful if there are many people running jobs on the remote host). The information inside the print statement tells you what part of the 'qstat' will be displayed on conky. 
            <p> [The alternative 'print' command below will show the status of the job (ex. 'R' for running), the running time of the job, and then the job name:] 
            <p style="font-family: 'Inconsolata', sans-serif;font-size: 13px">
            ${execi 10 ssh user@remotehost.edu 'qstat | grep username | gawk "{ print \$5 \": \" \$4 \"\t\" \$2 }"'}</p>
            
            <a name="myconky"></a>
            <h5> My .conkyrc file </h5>
            <p> If you want a configuration to get you started here is the whole text of my .conkyrc file (to get the qstat command to work fill in the needed information under the "Remote Host" section of the configuration) : </p>
            <?php include 'conkycode.php'; ?>
            
<p> Note: the Linux Mint 15 logo should be downloaded and saved.  Include the pathway of the logo under the "TEST" - "SYSTEM" tab "image /pathway/to/logo/...". <br /> In the end, your conky configuration should look something like this:
<br />
 <a href="http://oi40.tinypic.com/2emectc.jpg" target="_blank"><img src="http://oi40.tinypic.com/2emectc.jpg" width="720" border="0" alt="Conky"></a></p>
 
            <a name="nextconky"></a>
            <h5> What's next? </h5>
            <p> Further options include creating a RSA key that would allow you to bypass entering your password into the prompt conky gives when it logs into the remote host.  You could also customize your conky to display things like Twitter feeds or your music library. 

            
<h5><a style="color:#BA522B; font-size:16px" href="#conky">Top of "Using Conky"</a> <br /> <a style="color:#BA522B; font-size:16px" href="#top">Top</a></h5>

        <!Running novoalign>    
        <a name="novoalign"></a>    
        <h4>Running Novoalign:</h4>
        <ultxt2>
        <litxt2>|<a style="color: #BA522B" href="#backgroundnovo">Background</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#usenovo">How do we use it?</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#givenovo">What will Novoalign give you?</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#overallnovo">Overall</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#morenovo">More information</a>|</litxt2>
    </ultxt2>
        <!Background>
        
        <a name="backgroundnovo"></a>
        <h5>Background:</h5>
            <p>Novoalign is a software that aligns sequencing reads to a reference database(i.e.HG19) and calls variants to identify regions in the sequences that may be of significance. The software supports fragment, paired-end reads and is designed as an accurate and quick short read mapper.  Novoalign can be used for small to large genomes.</p>
            <p> Specifics: Novoalign 2.08.03 for Illumina paired-end reads.</p>
        <!How do we use it?>
        <a name="usenovo"></a>
        <h5>How do we use it?</h5>
            <p>Novoalign's maker - Novocraft - provides some documentation on how to use Novoalign. The website provided a formula to follow under their Quick Start Tutorial:</p> 
            
        <p style="font-family:'Inconsolata', sans-serif;font-size: 13px">           novoalign -d reference_genome -f filename1 filename2 -i mean, stdev -o SAM </p>
        
            <p> To run it, a reference genome had to be created.  The genome was created using these commands:</p>
            <p style="font-family:'Inconsolata', sans-serif;font-size: 13px">
            module load novocraft
            <br />novoindex ref.nix ref.fa
            </p>
            <p>An example of the command needed to actually run the aligner is:</p>
            <p style="font-family:'Inconsolata', sans-serif;font-size: 13px"> 
            novoalign -d ref.nix -f ../Sample/4_2_10_TAGCTT_L001_R1_001.fastq.gz ../Sample/4_2_10_TAGCTT_L001_R2_001.fastq.gz -i 250,90 -o SAM > L001_001.SAM </p>
            <p>(To calculate the mean and standard deviation of the size distribution of the sequencing runs, a short python program was created. However, you could also use this command: </p>
            <p style="font-family:'Inconsolata', sans-serif; font-size:13px"> novoalign -d genome.nix -f read1.fq read2.fq -i 500,100 -#2K >/some/output</p> 
            <p> This command will give the first 2000 alignments of two files, and the mean and standard deviation of the fragment lengths of the sequence will be displayed at the end of the file.  As long as the -i parameters are robust enough for the read [ex. -i 500,100], the alignment will be fairly accurate.)</p>
            
            <p> The .run file I created to run this program is below.</p>
            <?php include 'novoalignruncode.php'; ?>
    
    <!What do you get from it?>
    <a name="givenovo"></a>
        <h5>What will Novoalign give you?</h5>
        <p> You will get two output files: an *.oe file and a *.SAM file.  The *.oe file will give you:</p>
    <ul>
        <li> The time the run started and stopped </li>
        <li> Elapsed time in seconds, and CPU time in minutes </li>
        <li> The number of paired reads and pairs aligned, among other statistics</li>
        <li> Fragment length distribution (which is the length of the DNA fragment as mapped by the aligner) </li>
        <li> Mean and standard deviation of the fragment length of the DNA strands that were sequenced </li>
        </ul>
    <!Points about novoalign>
    <a name="overallnovo"></a>
        <h5>Overall:</h5>
        <p>Novoalign is one of the fastest alignment tools and has quite a high sensitivity for detection of variants. However, Novoalign requires relatively more memory than other alignment softwares (our run took about 8GB of memory).  It is also best used for an alignment of short reads to a long reference.  
        <!More information>
        <a name="morenovo"></a>
        <h5>More information:</h5>  
        <p> Novocraft's Website: <a href="http://www.novocraft.com/main/index.php" style="color: #80BCA3" target="_blank"> Main Page </a> </p>
    <p> Novoalign - Quick Start Tutorial: <a href="http://www.novocraft.com/wiki/tiki-index.php?page=Novoalign+NGS+Quick+Start+Tutorial&structure=Novocraft+Technologies&page_ref_id=76" style="color: #80BCA3" target="_blank"> Quick Start Tutorial </a> </p>
    <p> Novoalign - User Guide/Wiki: <a href="http://www.novocraft.com/wiki/tiki-index.php" style="color: #80BCA3" target="_blank"> User Guide</a> </p>
    <p> Novoalign - Reference Manual: <a href="http://www.novocraft.com/main/page.php?s=doc_novoalign_refmanual" style="color:#80BCA3" target="_blank"> Manual</a> </p> 
    <p> Novoalign - FAQ: <a href="http://www.novocraft.com/wiki/tiki-view_faq.php?faqId=1" style="color:#80BCA3" target="_blank"> FAQ </a> </p> 
    <h5><a style="color:#BA522B; font-size:16px" href="#novoalign">Top of "Running Novoalign"</a> <br /> <a style="color:#BA522B; font-size:16px" href="#top">Top</a></h5>     
    
    
<!beginning of chisq section>
    
    <a name="chisq"></a>    
        <h4>Running a Chi-Squared Test in Python:</h4>
        <ultxt2>
        <litxt2>|<a style="color: #BA522B" href="#overviewchisq">Overview</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#gwaschisq">What is a GWAS?</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#pythonchisq">Chi-Squared in Python</a>|</litxt2>
        <litxt2>|<a style="color: #BA522B" href="#codechisq">My Code</a>|</litxt2>
    </ultxt2>
        
        <!Overview>
        <a name="overviewchisq"></a>
        <h5>Overview:</h5>
        <p>A Chi-Squared test is statistical test that compares observed counts of categorical data with expected counts of the same categorical data.  To calculate the expected counts of a 2x2 contigency table you follow this formula:<br />
        <br />
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Expected = Row Total X Column Total / Total Counts <br />
        <br />
        To get the Chi-Squared value you follow this formula:<br />
        <br />
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chi-Squared = Î£ [(observed count - expected count)^2/expected count]<br />
        <br />
        Using the Chi-Squared value you can then measure statistical significance by calculating the p-value. The Chi-Squared test utilizes a null hypothesis, or the hypothesis that the two categorical variables are independent of each other. If the p-value is less than the assigned level of significance (in most cases, this value is 0.05), then you REJECT the null hypothesis and conclude that the two categorical variables are NOT independent.</p>
        <p>   
        This <a href="http://math.hws.edu/javamath/ryan/ChiSquare.html" style="color: #80BCA3" target="_blank"> link</a> provides a great and detailed explanation of the Chi-Squared test. </p>
        
        <!GWAS>
        <a name="gwaschisq"></a>
        <h5>What is a GWAS?</h5>
        <p>A Genome Wide Association Study (or GWAS for short) compares a 'control' genome (ex. HG19) with a 'case' genome (ex. your genome) and genotypes both samples for common single-nucleotide polymorhism's (or SNPs for short).  A GWAS aims to determine if allelic frequency of SNPs between the two genomes is different. </p>
        <p> To determine this, we do an odds ratio.  When an odds ratio is 1 it means that both SNPs have equal odds of happening.  A Chi-Squared is then used to test the significance of the odds ratio. </p>
        
        <a name="pythonchisq"></a>
        <h5>Chi-Squared in Python:</h5>
        <p> To start creating this Python file, I first had to figure out how to do a Chi-Squared test and understand how it works.  A link for a good summary of that can be found in the "Overview" section.  I further had to find a sample GWAS dataset that I could use for my script.  The PDF I found not only provided a sample GWAS dataset, but it also provided a great tutorial for understanding the GWAS dataset and using it in a program called PLINK.  Follow this <a href="https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&ved=0CC0QFjAA&url=http%3A%2F%2Fbioinf.wehi.edu.au%2Fbioinfosummer2010%2Fmaterials%2FBioInfoSummer2010_GWAS_tutorial.pdf&ei=qOQTUpXLBsSEiALJ3oGYDg&usg=AFQjCNEeuLVExxwQOu-G6DCKBmaNRDDLRQ&sig2=xRmPlfN4m8qdq2NCTvr6rw&bvm=bv.50952593,d.cGE" style="color:#80BCA3" target="_blank">link</a> to download the PDF.  If it does not download, try googling "BioInfoSummer2010 GWAS tutorial - WEHI Bioinformatics".
        <p> My script uses a 2x2 contigency table for single variants (1 pair of SNPs at a time).  My script uses 1 degree of freedom and outputs a p-value if it has less than 1.0e-8 significance, but both of these values can be changed in the script.  A future step in using this script would be to altering it so that it can run Chi-Squared association tests for multiple variants as well as single variants.</p>
        
        <a name="codechisq"></a>
        <h5>My Code:</h5>
        
    <?php include 'pythoncode.php'; ?> 
    <h5><a style="color:#BA522B; font-size:16px" href="#chisq">Top of "Running a Chi-Squared Test in Python"</a> <br /> <a style="color:#BA522B; font-size:16px" href="#top">Top</a></h5>
    <p style="border-style: dotted; border-width: 0px; border-top-width: 5px; margin-left: 0px; margin-right:0px"> </p>  
    

      
<!beginning of 'Useful Links' section>
    <a name="useful"></a>
    <h3>Useful Links: </h3>
    <p> The Scripps Translational Science Institute (STSI): <a href="http://www.stsiweb.org/" style="color: #80BCA3" target="_blank"> www.stsiweb.org</a> </p>
    <p> The Scripps Research Institute (TSRI): <a href="http://www.scripps.edu/" style="color: #80BCA3" target="_blank"> www.scripps.edu</a> </p>
    <p> Scripps Health: <a href="http://www.scripps.org/" style="color: #80BCA3" target="_blank"> www.scripps.org</a> </p>
    <p> Genomics and Current Projects - Dr. Carland (Main Page): <a href="http://genomics1.scripps.edu/" style="color: #80BCA3" target="_blank"> genomics1.scripps.edu</a> </p>
    <h5><a style="color:#BA522B; font-size:16px" href="#top">Top</a></h5>
    <p style="border-style: dotted; border-width: 0px; border-top-width: 5px; margin-left: 0px; margin-right:0px"> </p>
    
        
<!beginning of 'Contact Me' section>    
<a name="contact"></a>
<h3> Contact Me: </h3>
     <p> Email: <a href="mailto:mollee@scripps.edu" style="color:#80BCA3"> mollee@scripps.edu</a></p>
     <p> Secondary Email: <a href="mailto:mjain2@wellesley.edu" style="color:#80BCA3"> mjain2@wellesley.edu</a></p>
     <h5><a style="color:#BA522B; font-size:16px" href="#top">Top</a></h5>
     <br />
     <!p style="border-style: dotted; border-width: 0px; border-top-width: 5px; margin-left: 0px"> <!/p>
    
</body>
</html>
